In this project, I will be processing images with a CNN in a 2x2 experimental design. For this experiment I will use the train and test pattern with 10 epochs and a batch size of 64. I used 2 different combinations of activation functions (Sigmoid and Softmax) and optimizers (Adam and AdaDelta) and created 4 models. Each model was split into 80% Train and 20% Test dataset and the accuracy and loss was calculated on each dataset. The CNN model has 2 Convolutional layers, 2 Pooling layers, 1 Flatten and 2 Dense layers. All the images have been converted to 64 x 64 size and read as a Grayscale image.

Here are the Results from Top two Models:

1. Model 1 - Sigmoid activation function and Adam Optimizer - Final loss: 0.2458 - Train Accuracy: 0.9044 - val_loss: 1.0901 - Test Accuracy: 0.5925
2. Model 2 - Sigmoid activation function and AdaDelta Optimizer - Final loss: 0.4120 - Train Accuracy: 0.8188 - val_loss: 0.7452 - Test Accuracy: 0.6150

<b>Management Problem:</b>
If accuracy is a priority then I will recommend using a CNN model with Sigmoid activation function and AdaDelta optimizer for Image Classification as the difference between Train and Test Accuracies is less and Loss in Test Dataset is lesser than that of Model 1. The CNN model should have 2 Convolutional layers, 2 Pooling layers, 1 Flatten and 2 Dense layers. Image size 64 X 64 works best for this model as per the experiment so it should be used for building the image classifier.

## Importing Train and Test files for Cat and Dog dataset

In [6]:
# Importing the required packages to create CNN

import numpy as np
import pandas as pd
import cv2
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Dropout
from tensorflow.keras.layers import Activation, Conv2D, MaxPooling2D

import os

# Setting the path to cat dog image directory
path = "D:/Northwestern/DS 422/Assignment 7/cats_dogs_images/cats_dogs"

In [7]:
# Getting the images into arrays
# Dogs are categorized as 1 and Cats as 0

X = []
y = []
convert = lambda category : int(category == 'dog')
def create_test_data(path):
    for p in os.listdir(path):
        category = p.split(".")[0]
        category = convert(category)
        img_array = cv2.imread(os.path.join(path,p),cv2.IMREAD_GRAYSCALE)
        new_img_array = cv2.resize(img_array, dsize=(64, 64))
        X.append(new_img_array)
        y.append(category)

# Getting the images in an array and reshaping
create_test_data(path)
X = np.array(X).reshape(-1, 64,64,1)
y = np.array(y)

In [8]:
# Setting the seed
RANDOM_SEED = 999

# Normalize data
X = X/255.0

## Model 1 - Sigmoid activation function and Adam optimizer

In [9]:
# Creating a model with 2 Convolutional and 2 Pooling layers
# 1 Flatten and 2 Dense layers

model = Sequential()
# Adds a densely-connected layer with 64 units to the model:
model.add(Conv2D(64,(3,3), activation = 'relu', input_shape = X.shape[1:]))
model.add(MaxPooling2D(pool_size = (2,2)))
# Add another:
model.add(Conv2D(64,(3,3), activation = 'relu'))
model.add(MaxPooling2D(pool_size = (2,2)))

model.add(Flatten())
model.add(Dense(64, activation='relu'))
# Add a softmax layer with 10 output units:
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer="adam",
              loss='binary_crossentropy',
              metrics=['accuracy'])

## Experiment 1

In [10]:
# Fitting the model for 10 epochs
# Batch size 32
# Creating a validation split of 20%

model.fit(X, y, epochs=10, batch_size=32, validation_split=0.2)

Train on 1600 samples, validate on 400 samples
Epoch 1/10
1600/1600 [==============================] - 23s 15ms/sample - loss: 0.6828 - acc: 0.6081 - val_loss: 0.8321 - val_acc: 0.0000e+00
Epoch 2/10
1600/1600 [==============================] - 23s 15ms/sample - loss: 0.6663 - acc: 0.6250 - val_loss: 0.8515 - val_acc: 0.0000e+00
Epoch 3/10
1600/1600 [==============================] - 23s 14ms/sample - loss: 0.6415 - acc: 0.6381 - val_loss: 0.9871 - val_acc: 0.0500
Epoch 4/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 0.5985 - acc: 0.6850 - val_loss: 0.8491 - val_acc: 0.4200
Epoch 5/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 0.5448 - acc: 0.7181 - val_loss: 1.1537 - val_acc: 0.2700
Epoch 6/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 0.4844 - acc: 0.7731 - val_loss: 0.8858 - val_acc: 0.4825
Epoch 7/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 0.4410 - acc: 0.7956 - val_loss: 0.88

Final loss: 0.2458 - Train Accuracy: 0.9044 - val_loss: 1.0901 - Test Accuracy: 0.5925

## Model 2 - Sigmoid activation function and AdaDelta optimizer

In [11]:
# Creating a model with 2 Convolutional and 2 Pooling layers
# 1 Flatten and 2 Dense layers

model2 = Sequential()
# Adds a densely-connected layer with 64 units to the model:
model2.add(Conv2D(64,(3,3), activation = 'relu', input_shape = X.shape[1:]))
model2.add(MaxPooling2D(pool_size = (2,2)))
# Add another:
model2.add(Conv2D(64,(3,3), activation = 'relu'))
model2.add(MaxPooling2D(pool_size = (2,2)))

model2.add(Flatten())
model2.add(Dense(64, activation='relu'))
# Add a softmax layer with 10 output units:
model2.add(Dense(1, activation='sigmoid'))

model2.compile(optimizer="adadelta",
              loss='binary_crossentropy',
              metrics=['accuracy'])

## Experiment 2

In [12]:
# Fitting the model for 10 epochs
# Batch size 32
# Creating a validation split of 20%

model2.fit(X, y, epochs=10, batch_size=32, validation_split=0.2)

Train on 1600 samples, validate on 400 samples
Epoch 1/10
1600/1600 [==============================] - 23s 14ms/sample - loss: 0.6818 - acc: 0.6225 - val_loss: 1.1529 - val_acc: 0.0000e+00
Epoch 2/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 0.6648 - acc: 0.6250 - val_loss: 0.7885 - val_acc: 0.0075
Epoch 3/10
1600/1600 [==============================] - 23s 15ms/sample - loss: 0.6556 - acc: 0.6294 - val_loss: 0.8452 - val_acc: 0.0050
Epoch 4/10
1600/1600 [==============================] - 24s 15ms/sample - loss: 0.6328 - acc: 0.6444 - val_loss: 0.6855 - val_acc: 0.6900
Epoch 5/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 0.5972 - acc: 0.6637 - val_loss: 0.8146 - val_acc: 0.4300
Epoch 6/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 0.5906 - acc: 0.7038 - val_loss: 1.0114 - val_acc: 0.2625
Epoch 7/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 0.5238 - acc: 0.7506 - val_loss: 0.7903 -

Final loss: 0.4120 - Train Accuracy: 0.8188 - val_loss: 0.7452 - Test Accuracy: 0.6150

## Model 3 - Softmax activation function and Adam optimizer

In [13]:
# Creating a model with 2 Convolutional and 2 Pooling layers
# 1 Flatten and 2 Dense layers

model3 = Sequential()
# Adds a densely-connected layer with 64 units to the model:
model3.add(Conv2D(64,(3,3), activation = 'relu', input_shape = X.shape[1:]))
model3.add(MaxPooling2D(pool_size = (2,2)))
# Add another:
model3.add(Conv2D(64,(3,3), activation = 'relu'))
model3.add(MaxPooling2D(pool_size = (2,2)))

model3.add(Flatten())
model3.add(Dense(64, activation='relu'))
# Add a softmax layer with 10 output units:
model3.add(Dense(1, activation='softmax'))

model3.compile(optimizer="adam",
              loss='binary_crossentropy',
              metrics=['accuracy'])

## Experiment 3

In [14]:
# Fitting the model for 10 epochs
# Batch size 32
# Creating a validation split of 20%

model3.fit(X, y, epochs=10, batch_size=32, validation_split=0.2)

Train on 1600 samples, validate on 400 samples
Epoch 1/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 2/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 3/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 4/10
1600/1600 [==============================] - 23s 14ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 5/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 6/10
1600/1600 [==============================] - 22s 13ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 7/10
1600/1600 [==============================] - 23s 14ms/sample - loss: 9.9640 - acc: 0.3750 

Final loss: 9.9640 - Train Accuracy: 0.3750 - val_loss: 1.1921e-07 - Test Accuracy: 1.0000

## Model 4 - Softmax activation function and AdaDelta optimizer

In [15]:
# Creating a model with 2 Convolutional and 2 Pooling layers
# 1 Flatten and 2 Dense layers

model4 = Sequential()
# Adds a densely-connected layer with 64 units to the model:
model4.add(Conv2D(64,(3,3), activation = 'relu', input_shape = X.shape[1:]))
model4.add(MaxPooling2D(pool_size = (2,2)))
# Add another:
model4.add(Conv2D(64,(3,3), activation = 'relu'))
model4.add(MaxPooling2D(pool_size = (2,2)))

model4.add(Flatten())
model4.add(Dense(64, activation='relu'))
# Add a softmax layer with 10 output units:
model4.add(Dense(1, activation='softmax'))

model4.compile(optimizer="adadelta",
              loss='binary_crossentropy',
              metrics=['accuracy'])

## Experiment 4

In [16]:
# Fitting the model for 10 epochs
# Batch size 32
# Creating a validation split of 20%

model4.fit(X, y, epochs=10, batch_size=32, validation_split=0.2)

Train on 1600 samples, validate on 400 samples
Epoch 1/10
1600/1600 [==============================] - 24s 15ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 2/10
1600/1600 [==============================] - 24s 15ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 3/10
1600/1600 [==============================] - 22s 14ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 4/10
1600/1600 [==============================] - 23s 14ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 5/10
1600/1600 [==============================] - 24s 15ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 6/10
1600/1600 [==============================] - 23s 14ms/sample - loss: 9.9640 - acc: 0.3750 - val_loss: 1.1921e-07 - val_acc: 1.0000
Epoch 7/10
1600/1600 [==============================] - 23s 14ms/sample - loss: 9.9640 - acc: 0.3750 

Final loss: 9.9640 - Train Accuracy: 0.3750 - val_loss: 1.1921e-07 - Test Accuracy: 1.0000